app/src/ingestion/markdown_chunking.py

```diff
@@ -0,0 +1,341 @@
+            child_data_ids,
+            set(n.data_id for n in nodes_to_remove),
+        )
+    for child_node in nodes_to_remove:
+        remove_child(node, child_node)
+
+
+def find_closest_ancestor(
+    node: Node, does_match: Callable[[Node], bool], include_self: bool = False
+) -> Node:
+    "Return the first parent/ancestor node that matches the does_match function"
+    if include_self and does_match(node):
+        return node
+
+    while node.parent:
+        if does_match(node.parent):
+            return node.parent
+        node = node.parent
+    return None
+
+
+def nodes_as_markdown(nodes: Sequence[Node]) -> str:
+    # Don't render Heading nodes by themselves
+    if len(nodes) == 1 and nodes[0].data_type in ["Heading"]:
+        return ""
+
+    md_list: list[str] = []
+    for node in nodes:
+        # node.data["summary"] is set when node is chunked
+        node_md = node.data["summary"] or render_subtree_as_md(node, normalize=True)
+        md_list.append(node_md)
+    return normalize_markdown("".join(md_list))
+
+
+@dataclass
+class ProtoChunk:
+    "Temporary data structure for storing chunk data before creating a Chunk object"
+    id: str
+    headings: list[str]
+    markdown: str  # Markdown content of the chunk
+    # to_embed: str  # TODO: create string to embed from headings and markdown
+
+
+class ChunkingConfig:
+
+    def __init__(self, max_length: int) -> None:
+        self.max_length = max_length
+        self.chunks: dict[str, ProtoChunk] = {}
+
+    def fits_in_chunk(self, markdown: str) -> bool:
+        return len(markdown.split()) < self.max_length
+
+    def nodes_fit_in_chunk(self, nodes: Sequence[Node]) -> bool:
+        return self.fits_in_chunk(nodes_as_markdown(nodes))
+
+    def create_chunk(
+        self,
+        nodes: Sequence[Node],
+        chunk_id_suffix: Optional[str] = None,
+        breadcrumb_node: Optional[Node] = None,
+    ) -> ProtoChunk:
+        if not chunk_id_suffix:
+            chunk_id_suffix = nodes[0].data_id
+        chunk_id = f"{len(self.chunks)}:{chunk_id_suffix}"
+        chunk = ProtoChunk(
+            chunk_id,
+            get_parent_headings_raw(breadcrumb_node or nodes[0]),
+            nodes_as_markdown(nodes),
+        )
+        if not self.fits_in_chunk(chunk.markdown):
+            raise AssertionError(f"{chunk_id} is too large! Check before calling create_chunk()")
+        logger.info("Created chunk %s from %i nodes", chunk_id, len(nodes))
+        self.chunks[chunk.id] = chunk
+        return chunk
+
+    def create_chunks_for_next_nodes(self, node: Node, intro_node: Optional[Node] = None) -> None:
+        # TODO: Splitting the contents of the 2 nodes into chunks
+        raise AssertionError(
+            f"{node.parent.data_id}: These node(s) cannot fit into a single chunk:"
+            f" {node.data_id} {intro_node.data_id if intro_node else ''}"
+        )
+        # TODO: Add to self.chunks
+        # self.chunks[chunk.id] = chunk
```

Comment:
The need for this is very unlikely for realistic data and embedding model. Basically, this would only be called for a large paragraph that doesn't fit in a chunk. I'd noodling around with splitting to fill up the chunks -- e.g., if text requires 2 chunks, then `chunk[0]=text[:limit]` and `chunk[0]=text[-limit:]`, which maximizes the chunk size and chunk_overlap size.